### ISSUE TITLE
[bug] Incorrect TMS9x2x VDP family palette

### ISSUE THREAD

**Issue by sdsnatcher** [openMSX/openMSX#1024]
As is widely known, the TMS9x2x VDP family outputs its colors in YPbPr. OpenMSX currently uses for the TMS9x2x VDPs a palette obtained from Wikipedia.

While working on bug sf#306 (#607 on github), I researched about this topic and was able to produce a more precise palette using the following procedure, with the first step being very similar to the one used to reverse engineer the Toshiba T7937 palette:

1) Convert the YPbPr values from the datasheet to digital 8bit/channel DAC values in YCbCr
2) Use the official formula from the awarded BT.601 standard to convert the YCbCr to digital 24bit RGB

Check the attachment at the following address to see the results:
**http://damad.be/joost/openmsx/attach/p/openmsx/bugs/_discuss/thread/890f77af/fb98/attachment/TMS9928_family_Palette.xls**

Lastly, the TMS9x18 couple have built-in NTSC encoders that output CVBS, but I won't tackle those for now nor in this thread. It requires more research and I have the impression that this two VDPs might feature a different palette from their YPbPr counterparts. For now I think it's enough for openMSX to use the TMS9x2x palette for them too.

For more background details and references, please check #607 .

**Comment by MBilderbeek** [comment]
Some questions:
- from that sheet, you mean the table under 'Digital conversion" to be used in openMSX, right?
- for which VDP types should we apply this palette table? 
  We currently have the following flavours:

```
        enum VdpVersion {
                /** MSX1 VDP, NTSC version.
                  * TMS9918A has NTSC encoding built in,
                  * while TMS9928A has color difference output;
                  * in emulation there is no difference. */
                TMS99X8A   = VM_MSX1 | VM_PALCOL_MIRRORING | VM_VRAM_REMAPPING,

                /** MSX1 VDP, PAL version. */
                TMS9929A   = VM_MSX1 | VM_PALCOL_MIRRORING | VM_VRAM_REMAPPING | VM_PAL,

                /** newer variant PAL. */
                TMS9129    = VM_MSX1 | VM_PAL,

                /** newer variant NTSC. */
                TMS91X8    = VM_MSX1,

                /** Toshiba clone (hardwired as PAL). */
                T6950PAL   = VM_MSX1 | VM_TOSHIBA_PALETTE | VM_NO_MIRRORING | VM_PAL,

                /** Toshiba clone (hardwired as NTSC). */
                T6950NTSC  = VM_MSX1 | VM_TOSHIBA_PALETTE | VM_NO_MIRRORING,

                /** VDP in Toshiba T7937A engine (hardwired as PAL). */
                T7937APAL  = VM_MSX1 | VM_TOSHIBA_PALETTE | VM_PAL,

                /** VDP in Toshiba T7937A engine (hardwired as NTSC). */
                T7937ANTSC = VM_MSX1 | VM_TOSHIBA_PALETTE,

                /** MSX2 VDP. */
                V9938      = 0,

                /** MSX2+ and turbo R VDP. */
                V9958      = VM_YJK,
        };
```
- What should happen with the palettes of the MSX1 VDP's that do not have the pallette you're proposing in this ticket?

**Comment by sdsnatcher** [comment]
Yes, use the palette from the Digital Conversion box. Let's call it "TMS-BT.601 palette" from now on.

For now, use the TMS-BT.601 palette for the following VDPs:  TMS99X8A, TMS9929A, TMS9129, TMS91X8.

But more research seems to be required specifically for the TMS9918 and TMS9118. The colors on these two seem to have a more orange tint than the others. I'm not sure if this is caused by some quirk of the NTSC encoder, or if the palette values tweaked to look better on NTSC.   But since we don't have enough information about this yet, just use the TMS-BT.601 palette with them for now.

> What should happen with the palettes of the MSX1 VDP's that do not have the pallette you're proposing in this ticket?

Quick summary:

1) VDPs that we already know the palette for sure:
- TMS9928, TMS9929, TMS9128, TMS9129: Use the TMS-BT.601 palette
- T7937: Use the T7937 palette
- Fujitsu FM-X via RGB: Use the TMS-3bitRGB palette shown in #1025 

2) VDPs that seem to have different palettes, but require further research
- TMS9918, TMS9118: for now, closest approximate is to use the TMS-BT.601 palette
- T6950: for now, the closest approximate is to use the T7937 palette
- Fujitsu FM-X via CVBS: It seems to use same 16-color palette as the ZX-Spectrum. The closest approximate for now is the TMS-3bitRGB.

**Comment by MBilderbeek** [comment]
Got some more questions:
- Can you give the exact formula (or precise reference to it), which you used to calculate the RGB values?
- What exactly are these two auxiliary tables for?

**Comment by MBilderbeek** [comment]
New palette based on TMS-BT.601:
![New palette based on TMS-BT.601](https://cloud.githubusercontent.com/assets/12675978/15335903/6ad1ddd6-1c74-11e6-923a-5265352f34c0.png)
Original TMS9xxx:
![Original](https://cloud.githubusercontent.com/assets/12675978/15335904/6ad268fa-1c74-11e6-9de1-747e0dc8e16e.png)
T6950:
![T6950](https://cloud.githubusercontent.com/assets/12675978/15336162/747fc892-1c75-11e6-8142-ded09f9f9468.png)
V9938:
![V9938](https://cloud.githubusercontent.com/assets/12675978/15336163/749f87f4-1c75-11e6-9234-fb0f26017498.png)

**Comment by sdsnatcher** [comment]
- The formulas are contained in each of the excel cells. Just select a given cell to see its formula. They were obtained in the official BT.601 v7 document published by ITU. All references are cited in the Excel file itself.

The conversion took 3 steps. The result of each step is represented by one table in the top.

Step-1: Convert the analog levels from theTMS9928 datasheet (YPbPr) to digital DAC 5bit/channel YCbCr values  (the same process used on T7937 to obtain the digital DAC values)
Step-2: Convert the DAC 5bit/ch YCbCr to 8bit/ch YCrCb and normalize the values to ITU-R BT.601 mandatory ranges
Step-3: Convert the 8bit/ch YCbCr ITU-BT.601 compliant values into RGB using the BT.601 conversion formulas. The resulting table is the "Digital Conversion ITU-R BT.601 RGB" table.
- The auxiliary tables are used to make the intermediary steps open and easier to understand. I have included balloon comments for each relevant cell. You can stop the mouse pointer over those cells and read the comments.

**Comment by MBilderbeek** [comment]
Well, the thing is, I'd also like to document it properly in the source code, in text form. Perhaps you can improve that here... also did you take a look how we got to the current palette?

See https://github.com/openMSX/openMSX/blob/master/src/video/Renderer.cc
I'd like to make the text just as insightful and understandable as the one for the old palette.

How can we ever decide which palette is correct? Especially with all these encoder circuits changing the colors, differently for many machines...

**Comment by sdsnatcher** [comment]
The formulas are these:

R=255/219_(Y-16)+255/112_(1-Kr)_(Cr-128)
G=255/219_(Y-16)-255/112_(1-Kb)_Kb/Kg_(Cb-128)-255/112_(1-Kr)_Kr/Kg_(Cr-128)
B=255/219_(Y-16)+255/112_(1-Kb)*(Cb-128)

The results must be clipped between 0 and 255.

Kr = 0,299
Kg = 0,587
Kb = 0,114

Wikipedia has a simplified version of the formula:
https://en.wikipedia.org/wiki/YCbCr#ITU-R_BT.601_conversion

And, who would have thought?! While checking the formulas to explain it here, I found a typo in the spreadsheet, in the green formula!! I fixed it, and the palette values have changed.

The interesting part is that the digital conversion results now have become very close to the results of the "El Cheapo RGB" formula! 

I attached the corrected spreadsheet here.

[TMS9928_family_Palette.xlsx](https://github.com/openMSX/openMSX/files/273681/TMS9928_family_Palette.xlsx)

**Comment by MBilderbeek** [comment]
Thanks, but can you also give the used formulas to get to the Y, Cb, Cr values you used?

Here's the corrected palette in openMSX in the same series as before:
New palette based on TMS-BT.601:
![openmsx2357](https://cloud.githubusercontent.com/assets/12675978/15438608/4d72cfe2-1ecd-11e6-81fc-3d39adcea72a.png)
Original TMS9xxx:
![Original](https://cloud.githubusercontent.com/assets/12675978/15335904/6ad268fa-1c74-11e6-9de1-747e0dc8e16e.png)
T6950:
![T6950](https://cloud.githubusercontent.com/assets/12675978/15336162/747fc892-1c75-11e6-8142-ded09f9f9468.png)
V9938:
![V9938](https://cloud.githubusercontent.com/assets/12675978/15336163/749f87f4-1c75-11e6-9234-fb0f26017498.png)

**Comment by sdsnatcher** [comment]
If you had asked all these questions two years ago you would have done my life much easier, you know?    :)

Give me some time to answer this one. I need to relearn the whole thing again.

**Comment by sdsnatcher** [comment]
Well, I found another error, but with a good reason. It was caused because back then I had no TMS9118 to check the real colors against. Remember when I said that I had to check the real TMS9118 (or TMS9918)? Well, since then I bought a Japanese Sony HB-10 for my tests, and this machine has the TMS9118 inside.

Back then, I was checking the results against the video output of my Sharp Hotbit, that has the LM1889 TV-Encoder inside. Comparing with its output, I assumed that the difference was caused because of the lack of Gamma correction, then applied it in the YCbCr 5bit/ch to YCbCr 8bit/ch. And in fact this made the colors match the Hotbit output, and as we just have seen it also made it closely match the "El Cheapo RGB" method.

But now, comparing to the TMS9118 built-in encoder, it seems that the LM1889 desaturates the colors on the output considerably. (more about that later in this message)

Now I decided to remove the gamma correction and just scale the YCrCb 5bit/ch to BT.601 YCbCr 8bit/ch directly. And the result is that the colors are much more saturated and closer to the ones shown by TMS9118. The minimal differences can be explained by the effect that the NTSC color encoding has on the colors.

The new formulas for the "8bit/ch BT.601" table are:

Y=ROUND(Y5/Max5_(MaxY_BT601-MinY_BT601)+MinY_BT601)
Cr=ROUND(Cr5/Max5_(MaxCc_BT601-MinCc_BT601)+MinCc_BT601)
Cb=ROUND(Cb5/Max5*(MaxCc_BT601-MinCc_BT601)+MinCc_BT601)

Maybe you won't recognize this palette instantly, because no European machine had a TMS9x18 inside and these two VDPs are the only ones with an original Texas TV-Encoder inside. Because of the PAL color encoding, all European machines (and Brazilian too) will have inside a TMS9x2x VDP coupled with some external TV-Encoder, like the LM1889, the LVA510 or even the awful MC1377. Because of this, all of them will feature some kind of color distortion. 

The palette seen by PAL machines (and the Sharp Hotbit) is closer to the "Analog Component Video to RGB" table, because their TV-Encoders apply some form of gamma correction to the chroma values very similar to the one I did in the previous version of the spreadsheet.

OHOT, the Machines that have RGB output or MC1377 TV-Encoder will show a palette closer to the "El Cheapo RGB"  or the "Lazy El Cheapo RGB" tables, depending on the build quality of the machine.

I think the color difference mystery between the TMS9x18 and the TMS9x2x is now solved. They have the very same palette, but the TV-Encoder will distort the colors. The only way to see the real colors on a TMS9x2x would be to connect its output directly to a modern TV with component-video input. Or use a more modern TV-Encoder.

I attached here the latest version of the spreadsheet.

[TMS9928_family_Palette_v2.xlsx](https://github.com/openMSX/openMSX/files/275509/TMS9928_family_Palette_v2.xlsx)

**Comment by sdsnatcher** [comment]
I uploaded the same image (Malaika's Intro screen) so you can compare the palette differences between the TMS9118 built-in TV-Encoder and the results of a TV-Encoder that use the "Analog Component Video" method. Both pictures were taken from the very same CRT TV.

I have chosen this screen because it shows all the colors (I modified this screen a bit to add the light yellow in the "i" letter), and the authors have made excellent use of the palette.

Sony HB-10 (TMS9118) image:
![sony_hb-10](https://cloud.githubusercontent.com/assets/13080763/15443922/5fa7727c-1ec2-11e6-8985-6e8c281afd6f.JPG)

Sharp HB-8000 v1.2 (TMS9128+LM1889)
![sharp_hb-8000](https://cloud.githubusercontent.com/assets/13080763/15443974/b7d40f00-1ec2-11e6-8c4f-6b3cbaa846d4.JPG)

Note: Keep in mind that the camera automatically does some white balance that desaturate the colors a bit when compared to the naked eye.

Then, here are the palettes from the new version of the spreadsheet, so you can compare the results with the real machines:

Corrected BT.601
![malaika_bt601_640x480](https://cloud.githubusercontent.com/assets/13080763/15444109/cd6d6fb8-1ec3-11e6-8220-213e40954b87.png)

Analog Component Video formula:
![malaika_componentvideo_640x480](https://cloud.githubusercontent.com/assets/13080763/15444113/d679cab6-1ec3-11e6-892d-a5ce34ffc70d.png)

**Comment by MBilderbeek** [comment]
So, can you give a conclusion on what you recommend we implement for now? So, for which VDP types should we apply which palette table?

**Comment by MBilderbeek** [comment]
Here's the TMS-BTU palette with your latest spreadsheet in the same context as before:
![openmsx2358](https://cloud.githubusercontent.com/assets/12675978/15448373/a107cc14-1f60-11e6-986c-219490f989fe.png)
Original TMS9xxx:
![Original](https://cloud.githubusercontent.com/assets/12675978/15335904/6ad268fa-1c74-11e6-9de1-747e0dc8e16e.png)
T6950:
![T6950](https://cloud.githubusercontent.com/assets/12675978/15336162/747fc892-1c75-11e6-8142-ded09f9f9468.png)
V9938:
![V9938](https://cloud.githubusercontent.com/assets/12675978/15336163/749f87f4-1c75-11e6-9234-fb0f26017498.png)

To be honest, that new BTU palette seems way to bright (not enough washed out) compared to my memories of the MSX1. But that's just a first impression.

**Comment by sdsnatcher** [comment]
> To be honest, that new BTU palette seems way to bright (not enough washed out) compared
> to my memories of the MSX1. But that's just a first impression.

Yes, this is why I I mentioned that probably people from Europe wouldn't recognize this palette as the one in their fond memories.

This is because the Texas built-in encoder outputs the colors with much higher saturation than the other encoders. Or, if you prefer, the other encoders output much less saturated colors than the Texas built-in encoder.     ;)

By "saturation" here I mean the equivalent of that color saturation knob that every TV has. If you crank up the color knob of your TV up to 100% for a typical European MSX1, you will get nearly the same colors that the Texas encoder produces when the color knob is set to 50% in the same TV.

Given that, my conclusion and recommendation is that, for the VDPs that have their internal digital palette encoded in YCbCr, the TV-Encoder (or RGB-encoder) circuitry is relevant, as it causes significant differences in the user experience that must me taken into the consideration. This is the case of all VDPs from Texas Instruments.

OTOH, for the VDPs have their internal digital palette encoded in RGB (V9938, V9958, T6950 and T7937), the TV-Encoder circuit has no significant effect in the user experience, thus the current openMSX approach is sufficient.

Another conclusion I came up during this research is that the current Wikipedia's MSX1 palette is idealized. It was manually tweaked to produce perfect blacks and white. Given how the TMS YCbCr palette is unbalanced, no MSX1 machine will produce such colors.

That said, and given the conclusion that the Encoder is relevant to the user experience, I think it should be added as a parameter of the Texa's VDPs in the machine XML file, allowing the appropriate palette to be set for each machine, like this:

```
<VDP id="VDP">
  <version>TMS99X8A</version>
  <encoder>LVA510</encoder>
  <io base="0x98" num="2"/>
</VDP>
```

But I advise to keep the number of palette variations in a sane level. This is also intended to make it easier for the reverse engineers and developers to identify the correct palette with just some pictures of the screen and the PCBs.

Those are the cues to look to identify which palette the machine with a Texas Instruments VDP has:
- The VDP is a TMS9x18: BT.601 palette
- The machine has a RGB-out connector and the circuit has no trimpots for RGB calibration: "El Cheapo RGB" palette
- The machine has a RGB-out connector and the circuit has trimpots for RGB calibration: "El Cheapo RGB with trimpots" palette
- The VDP is a TMS9x2x with any other TV-Encoder: YPbPr to RGB palette (called "Analog component-video to RGB" in the spreadsheet)
- Fujitsu FM-X: 3bit-RGB palette

**Comment by sdsnatcher** [comment]
I know this difference in saturation (thus the perceived palette) for PAL (& derivatives) machines seem weird, but don't take just my word for it.  Connect yourselves any MSX1 that has an RGB output to your SCART TV/monitor and check the palette.   :) 

The palette will look like the YPrPb conversion (called "Analog component-video" table in the spreadsheet) with the saturation set somewhere between 90% to 100%.

This explanation brings me an idea for another possible solution for this palette issue: Always use the analog YPbPr conversion formula, and have a parameter in the XML file to allow the saturation for that machine to be set. openMSX would then calculate and set the palette of the machine boot.

This would work in a very similar way to the audio mixer levels already implemented in openMSX.

In example:

For the Gradiente Expert XP-800:

```
<VDP id="VDP">
  <version>TMS9128</version>
  <saturation>90%</saturation>
  <io base="0x98" num="2"/>
</VDP>

```

For the Sharp HB-8000:

```
<VDP id="VDP">
  <version>TMS9128</version>
  <saturation>50%</saturation>
  <io base="0x98" num="2"/>
</VDP>

```

For the Philips VG-8020:

```
<VDP id="VDP">
  <version>TMS9129</version>
  <saturation>50%</saturation>
  <io base="0x98" num="2"/>
</VDP>

```

For the Sony HB-10:

```
<VDP id="VDP">
  <version>TMS9118</version>
  <saturation>100%</saturation>
  <io base="0x98" num="2"/>
</VDP>

```

This solution seems more flexible and easier to implement.

**Comment by MBilderbeek** [comment]
For your reference, the currently used palette has this commit message:

```
commit 07c320643dbce9b9a3e7e0fa013c683cb5df35a1
Author: Wouter Vermaelen <m9710797@users.sourceforge.net>
Date:   Sat Dec 1 16:49:52 2012 +0000

    Improve TMS99X8/TMS9929 color palette

    This change was based on hit9918's request in this forum thread:
      http://www.msx.org/forum/msx-talk/openmsx/minor-openmsx-other-emulators-vdp-emulation-feature-missing?page=0
    (see hit9918's post of 28-11-2012, 21:30 and later).

    See comments in the code for more details.
```

So see https://www.msx.org/forum/msx-talk/openmsx/minor-openmsx-other-emulators-vdp-emulation-feature-missing

**Comment by MBilderbeek** [comment]
Good question from Wouter: <wouterv> i was just wondering if playing with saturation alone is enough to mimic the alternative palette FRS wants

**Comment by MBilderbeek** [comment]
Idea from hap:

```
<hap> hue change here for example http://smg.photobucket.com/user/Evilweredragon/media/nescolor.jpg.html
<Quibus> hap: hue change due to what?
<hap> tv
<hap> "I took this picture a long time ago, and it always was weird to me. The NES on two TV's, a CRT HDTV, and an old Commodore 128 monitor, showed the sky in Mario completely different. One was blue, and one was purple.
<hap> Both TV's were tinted correctly using a blue filter on video essentials, yet they show a completely different color.
<hap> "
<Quibus> The purple looking TV was wrong, I guess :P
<hap> nope
<Quibus> I don't see that text with it though
<n_n> i keep telling you there is no definite palette
<n_n> [22:14] <hap> neither are wrong
<n_n> for machines (NES,MSX1,etc) that don't output rgb directly
<Quibus> I don't think the sky was intended to be purple
<n_n> yes it was, in USA NTSC, sky is purple tint
<n_n> but not on everyone's tv
<n_n> and in europe PAL it's more common on the blue side
<n_n> what you'd need is an approximation, and then let the user change hue,saturation,contrast,etc
```

Perhaps that's the best idea? Tune it in video settings like your old TV/monitor?

**Comment by sdsnatcher** [comment]
Yes, I remember hi9918’s arguments on that thread. Back then I totally agreed with it. The current Wikipedia palette certainly makes things look better and well balanced.

But since I investigated further the palettes from the T7937 and the TMS VDPs, I started wondering: do we want an idealized emulation, or do we want it to be as closest as possible of the real machine, even on the aspects that weren’t the best ones?

After all, the TMS VDPs were well known for having a low-contrast palette that was hard to master back then. But the Japanese game designers turned this into an advantage.

And, palette wise, the V9938 was created to solve that issue too. They even tweaked the default palette to increase the contrast and have at least one pure tone of red, blue and green colors.

Yes, it’s indeed nice to have an ideal palette and that’s why I created the CoolColors. Anyone who wants an ideal palette is free to create his own for the MSX2. Do we really need to tweak the MSX1 palette for that?

Leaving the “philosophical" questions aside, now back on the technical aspects:       ;)

I’m aware of the NES palette emulation discussion, and I guess that what's might be missing in the NES community to get its definitive palette is to reverse engineer the internal digital DAC values as I did for the TMS9918. Without that, there will be [endless discussions on how the colors should look, based entirely on subjective personal tastes, without a good technical base to start with](http://board.zsnes.com/phpBB3/viewtopic.php?f=10&t=3298).

For the MSX, we now know that it all comes from that 15bit YCbCr digital palette, shown in the spreadsheet, that was deduced from the palette shown in the datasheet. Anything that changes this palette before it reaches the user eyes are after effects.

Now, the openMSX need to decide in how much detail they want to emulate the distortions that occur in the signal path that will produce a perceived palette for the user.

For those “after effects”, I would like to add some premises and some facts:

1) Unlike the NES, the TMS9x18 seems to feature a very good NTSC encoder, that does a pretty decent job in outputting consistent colors between different TV sets, and pretty consistent with the BT.601 standard too.

2) Unlike the NES, there are a plethora of very different MSX designs. Except for NES clones that no one would care to check the palette, original NES machines are pretty much clones of one another with the major difference in the palette being the three regions: JP, US and EU. 

3) PAL is pretty much immune to hue changes by design. In a poor PAL decoder, phase changes are shown as Hanover bars. In a decent PAL decoder with line buffer, phase changes will cause more desaturation. Since the MSX is directly connected to the TV via CVBS (and not being broadcast OTA), the phase changes are minimal.

4) The relevant stages of the video signal path are:
- VDP (digital palette and DAC)
- Encoder (either TV-Encoder or RGB-Encoder)
- TV video system idiosyncrasies/artifacts: NTSC/PAL bandwidth restrictions for either CVBS or S-Video.

5) The TV-Encoder is always a chip that has its own conversion matrix built-in. This has a strong effect in the color saturation and, depending on the formula this might have some effect in their hue.

6) The RGB-Encoders, OTOH, are always built with discrete components. The quality of the conversion varies. 

7) When speaking of analog TV color systems, "hue" and "luminosity" are not literally those, but in fact [tint](https://en.wikipedia.org/wiki/Tint_control) and [luma](https://en.wikipedia.org/wiki/Luma_%28video%29) respectively. The difference seems subtle but is important. This means that if you intend to implement "hue" distortions, you must use the tint formulas instead.

---

Then I would like to add the following personal considerations:

1) The majority of Japanese games seem to have been developed for the TMS9x18 palette. The fact that, when run on an MSX2, the Konami Game Collections set a palette very close to the BT601 seems to corroborate with this idea. OTOH, European titles (except speccy ports) seem to have been developed for the much less saturated palette shown by the external TV encoders.

2) The plethora of different MSX designs makes it too difficult to emulate so many idiosyncrasies, inaccuracies. I think the best option is to stick to the correct formulas and calculate the correct palette using them. Just like openMSX doesn’t try to emulate the inaccuracies of the machines that have an RGB palette.

---

That said, I personally think that the best cost/benefit solution seem to be to try to mimic the following stages of the video signal path:

1) TV-Encoder/RGB encoder: this can be easily implemented by having the palette stored in YPbPr and generating the RGB palette only on boot, based one parameter supplied in the XML file: saturation. 

2) CVBS/S-Video artifacts (tint changes included): this can be done with a renderer, just like the existing renderers. Personally I would recommend the [Blargg's NTSC libraries](http://slack.net/~ant/libs/ntsc.html). Tint and saturation can be further adjusted by this renderer, just like in a TV. This additional renderer can be included at some point in the future. There's no need rush it now.

The reasons why I recommend to have it implemented like this are:

1) Compatibility with code made for the RGB VDPs, since everything would still be done in the RGB domain. The same renderers could be used for both the TMS, Yamaha and Toshiba VDPs.

2) Simplicity/flexibility, as explained in the previous messages

3) Speed: no performance impact for slower machines. Lighter renderers can still be used for those.

**Comment by MBilderbeek** [comment]
Thanks for the long and detailed explanation. I agree with the approach in the last part of your message.

However, I _still_ have some questions.... sorry...
1. Do you mean we should use that YPbPr2RGB using saturation for all TMS VDP's in the last proposal? (And thus not use that BT.601 stuff.)
2. Having the palette in YPbPr means the table you mention as "Analog Component Video" in your latest spreadsheet, right? (You wrote YPBPr, but the table says YPrPb, I hope nothing is mixed up here.)
3. How should someone who is writing the hardware config file know what saturation percentage to use? We should make it easy for a user to write a config file if he has the schematic or a photo of the PCB to see the chips used. You can't read the saturation from any of this, so how could he know?
4. What is the formula to calculate from that YPbPr table into RGB with only 'saturation' as input parameter?

**Comment by sdsnatcher** [comment]
1. The YPbPr formula produces results very close to the BT.601 formula if the saturation is set to 100%. Try for yourself to check the results.

I'm only recommending the YPbPr formula because saturation control is a requisite for openMSX and the BT.601 formula, being designed for digital conversion, doesn't allow that. Otherwise the BT.601 formula would be the best one.
1. Yes, its the "Analog Component Video". YPrPb is a synonym for YPbPr. Just the order of the components is change, similar to how the RBG order used for the V9938 palette registers is still RGB. I just used it in that sequence because it was easier to check the formulas when I was creating the spreadsheet, as the Pr and Pb components get in the same sequence as in RGB.
2. The best way to adjust the saturation is to compare the openMSX side-by-side with the real machine. If the machine has only composite output, start with 50% and adjust it up or down accordingly. If the machine has RGB output, start with 100% and reduce it a little if it is too saturated.
   Taking photos of the MSX screen will not help in this case. The majority of the cameras do some or other form of white balance/image "enhancement"  that change the hue and the saturation of the image. But I think that a capture of the machine's video output (running that COLORBAR.BAS utility) using a video grabber should be enough for this task.
   After some time, you'll already have a good database of the saturation that each TV-Encoder produces. So, in example, all machines with the LVA510 encoder will probably have the same image saturation.

For the existing XML files, I would recommend to set following saturation values:
- Any MSX with a TMS9x18 VDP: 100%
- Sharp Hotbit and other machines with the LM1889 TV encoder: 50%
- Gradiente Expert (gray models): 95%
- Other machines with a TMS9x2x VDP and RGB output: 100%, until someone reports a better value
- Other machines with a TMS9x2x VDP and CVBS only output: 50%, until someone reports a better value
1. The formulas are:

```
R=Y+1,402*Pr*S
G=Y-0,344*Pb*S-0,714*Pr*S
B=Y+1,772*Pb*S
```

S: Saturation  (0.00=0% to 1.00=100%)
R,G,B must be clipped in the 0~255 range

Y  must be in the 0~255 range and Pb and Pr in the -128 ~ +127 range, so they are calculated from the TMS datasheet values with the following formulas:

```
Y=TMS_Y*255
Pb=TMS_RY*255-128
Pb=TMS_BY*255-128
```

TMS_Y: TMS9918's Y value
TMS_RY: TMS9918's R-Y value
TMS_RB: TMS9918's B-Y value

I just noticed that the YPbPr formulas are nearly the same as the one you had documented in the Renderer.hh file. But now:

1) There's no manual tweaking
2) Saturation will be set in the XML file, instead of being hardcoded. RGB values will be calculated on boot instead of being hardcoded too.

If you set the saturation in the XML file to 54%, you should obtain a palette that is very close to the one that openMSX was using until now.

**Comment by MBilderbeek** [comment]
OK, clear, but I'm not very happy with a method that means that users with rare machines have to tweak with openMSX saturation values to find the right value for their machine. Isn't there the VDP type in combination with a chipset (like LM1889) in combination with the kind of output connector (RGB/CVBS) that can easily be seen from the motherboard/schematic (and/or the connectors on the back of the machine) that defines which saturation should be used?

**Comment by sdsnatcher** [comment]
Before answering your question, I must add some extra information I just found:

Looking at the Hotbit schematics, it doesn't seem to try to change the saturation in any relevant way. Except for two pull-down resistors and their respective bypass capacitors, the R-Y and B-Y VDP outputs are pretty much directly connected to the LM1889 inputs.

But when I looked the Philips VG-8020 service manual, what I saw there made me want to implement separate the Pr and Pb saturations. Because the VG-8020 clearly has a tweak to boost the Pr level.

---

Now answering your question:

But think about it: doesn't exactly the same situation happen with the audio mixing parameters of the XML file?  (I mean the PSG/OPLL/etc volumes)     ;)

Yes, we can take the TV-Encoder type as a good starting point, and probably many machines with the same TV-Encoder will feature very similar saturation levels.

Well, if you want a more precise method, IMHO the best way would be to write an utility to analyze an image (*1) of the machine's output running the colorbar.bas app, to calculate how much the saturation deviates from the reference BT.601 image.

*1: The image must be captured by the owner using a video grabber

The formulas to calculate the saturations are:

SatPb=(B_0,5-R_10,69-G_0,331)/RPb
SatPr=(R_0,5-G_0,419-B_0,081)/RPr

RPb=Reference BT.601 Pb level for that color
RPr=Reference BT.601 Pr level for that color

The final SatPb and SatPr should be the respective average of all SatPb and SatPr obtained from each color.

This method should output a very faithful reproduction of the the colors of a given machine.

The XML file will have then to contain two saturation parameters (SatPb and SatPr) and the RGB formulas must be slightly enhanced:

```
R=Y+1,402*Pr*SatPr
G=Y-0,344*Pb*SatPb-0,714*Pr*SatPr
B=Y+1,772*Pb*SatPb
```

References:
http://www.equasys.de/colorconversion.html

**Comment by MBilderbeek** [comment]
It might be similar to the volumes in the XML, but these are currently hardly tweaked at all. We just copy paste them from another machine. One time the volumes were tweaked (for the CX5M series).

I know it would be possible to calculate it all back, but hardly anyone has a TV capturing device. So it's no point to start relying on that, I think. I'd rather make a good approximation by using the information we can get from PCB photos or schematic.
In that line, is it possible to only mention the encoder and output type to derive the saturation that should internally be used to calculate the proper palette? If yes, do you have the data of the combinations?

**Comment by sdsnatcher** [comment]
Even if doesn't get tweaked too often like the sound volumes, the approach is both simpler to implement and more flexible. Having a table coded inside openMSX with the saturation for each TV-Encoder is certainly more complicated, more restricted and harder to maintain, isn't it?

Such table could be just kept as a TXT file, or just comments in the renderer.hh file.

> do you have the data of the combinations?

The best data I have until now is the table I suggested some messages ago. They are based on visual comparisons with three machines I have here:
- Any MSX with a TMS9x18 VDP: SatPr=SatPb=100%
- Sharp Hotbit and (and probably other machines with the LM1889 TV encoder): SatPr=SatPb=50%
- Gradiente Expert (gray models): SatPr=SatPb=95%

And now, by looking at the VG-8020 service manual, my educated guess about its values would be:
- Philips VG-8020: SatPr=75%, SatPb=50%

For all other machines with no known data, I suggest to use:
- Other machines with a TMS9x2x VDP and RGB output: SatPr=SatPr=100%, until someone reports a better value
- Other machines with a TMS9x2x VDP and CVBS only output: SatPr=SatPb=54%, until someone reports a better value

Anyway, usb video grabbers are pretty cheap these days. I bought this one for myself in February, but since things take ages to arrive in Brazil, it still haven't. Some other items I bought one month earlier have just arrived, so I expecting it just for the next month.

http://www.ebay.com/itm/USB-2-0-YPbPr-Component-RCA-Composite-AV-S-Video-Video-Capture-Card-HD-Grabber-/191627544784?hash=item2c9de448d0

**Comment by MBilderbeek** [comment]
OK, Wouter also suggests to go for the saturation parameters.

But first let me try to understand now the details of your spreadsheet.

Step 1: the origin of the palette

The origin is still the voltages from the datasheet as was in our original palette. You added a conversion to digital values, but you do not use that for the "Analog Component Video" table... Correct? (Wasn't that the whole starting point for you anyway?)

Step 2: converting to component video

If you leave out the scaling to values between 0 and 255 (which is a good thing, I think), you just do
Y = Y
Pb = (B-Y)-0.5
Pr = (R-Y)-0.5

whereas the calculation of the currently used palette seems to subtract 0.47 and then scale the result between -0.5 and 0.5 (by dividing both Pb and Pr by 0.53 and multiplying by 0.5). This is indeed what you wrote before: keep black totally black and white totally white!

So far, this seems to be the only difference so far, right? Which one is correct, though?

Step 3: converting to RGB

And next (for both old and new methods), this value is then scaled with the saturation (0.54 fixed in the existing code, the parameters in your method) and converted into RGB with the formula you mentioned before (and is also in the code). This value is then _scaled_ into the 0-255 range...
I wondering if scaling is correct here. What would be the best way to map this into the 0-255 range? Should there be clipping instead of scaling?

With your methods and 100% saturation, this gives RGB values in the range -0.05-1.27. Clip and scale? Or shift and scale?
With the existing methods, and 100% saturation, this gives R-G-B values in the range 0-1.28. Just scale? (Black will be black, but white will not be white!)

With 54%: your methods gives RGB values in the range -0.03-1.02 (again: clip and scale? or shift and scale?), existing methods 0-1.00 (where white is white!), which seems easy to just scale...

If we change the saturation, should we scale differently for each saturation in order to get to the full 0-255 range? Or is there one scaling/clipping method (based on 54% or on 100%, or another percentage saturation?) for all saturations?

**Comment by sdsnatcher** [comment]
> Step 1: the origin of the palette

Yes, the digital and analog conversions of the spreadsheet are independent.
- The analog conversion works over the analog voltage levels from the datasheet
- The digital conversion works over the DAC digital values

The only point in common is that the DAC digital values where deduced from the analog levels.

> Step 2: converting to component video
> If you leave out the scaling to values between 0 and 255 (which is a good thing, I think), you just do

I'm not sure if leaving the scaling out is a good thing. The color conversion formulas expect Pb and Pr to be within the -127.5 to +127.5 range, and Y to be in the 0~255 range.. See [here](http://www.equasys.de/colorconversion.html).

The best formula then would be:

```
Y = Y*255
Pb = ((B-Y)-0.5)*255
Pr = ((R-Y)-0.5)*255
```

> the currently used palette seems to subtract 0.47 and then scale the result between -0.5 and 0.5 (by dividing both Pb and Pr by 0.53 and multiplying by 0.5).

This is the step where the tweaking starts. This step tries to compensate the lack of precision of the TMS DAC (thus cannot produce true whites and blacks) by doing a shift in its scale. It's the equivalent to saying "Oh, I would love to have an intermediary tone of red between the V9938 (7,0,0) and the (6,0,0) levels, so let me just tweak the scale to my desire"     ;)

Please keep in mind that it wouldn't be possible to do such precise tweaking in the 80's home analog circuitry without _very expensive_ professional video equipment. What would be possible to do was:

1) To increase the _entire_ scale linearly by 6.38% to get the 0.47V adjusted to 0.5V, for example.
2) To add some value (positive or negative) to the Pb and Pr scales. This is what the TINT control does

I've been thinking and it's possible to emulate this kind of tweaking in the palette conversion. It will require one extra parameter in the conversion formulas, that are two extra saturation parameters applied to B-Y and R-Y before they're fed to the TV-Encoder. Yes, this will allow even greater control over the resulting palette, but then again:
- Remember my argument about the plethora of different machines and the decision to keep the formulas "by the book"?
- It will be very hard to reverse engineer these pré-TV-encoder saturation values to calibrate the conversion properly

If you really want that, I can add these extra saturation parameters to the formulas.

> So far, this seems to be the only difference so far, right? Which one is correct, though?

IMHO, the correct is to leave the formulas and the Analog Levels table all by-the-book. Some level of  tweaking can be done by the two saturation levels contained in the XML file. 

I'm not sure if you're aware, but the two Saturation parameters implemented in the formula are meant to emulate the TV-Encoder internal color matrix.

Any tweaks done to the R-Y and B-Y levels prior to their entering in the TV-Encoder/RGB-Encoder are not implemented in the formula. If you want me to implement those, I'll add them. This way you'll be able to tweak the TMS levels to obtain the pure white and blacks. But we agreed that this would be better to be done in the TINT control of the rendered, didn't we?

Step 3: converting to RGB

In the new method, the scaling is done as the first step, not the last. The YPbPr formulas will then spit the RGB values in the RGB scale, but they must be clipped. It's very similar to the YJK->RGB conversion on the MSX, since both are based in exact the same principles.

In a nutshell, these are the steps that must be done in the new method, in this exact sequence:

Step-1: Scale the TMS Analog Values to the YPbPr formula expected ranges

```
Y = TMS_Y*255
Pb = (TMS_BY-0.5)*255
Pr = (TMS_RY-0.5)*255
```

Step-2: Convert YPbPr to RGB using the standard formulas with saturation:

```
R=Y+1,402*Pr*SatPr
G=Y-0,344*Pb*SatPb-0,714*Pr*SatPr
B=Y+1,772*Pb*SatPb
```

Step-3: Clip the resulting RGB values to the 0~255 range

```
R=CLIP(R,0,255)
G=CLIP(G,0,255)
B=CLIP(B,0,255)
```

**Comment by MBilderbeek** [comment]
Taking it one step at a time...

> I'm not sure if leaving the scaling out is a good thing. The color conversion formulas expect Pb and Pr to be within the -127.5 to +127.5 range, and Y to be in the 0~255 range.. See here.
> 
> The best formula then would be:
> 
> Y = Y_255
> Pb = ((B-Y)-0.5)_255
> Pr = ((R-Y)-0.5)*255

That extra 255 doesn't matter, it's just an overall factor you can leave out. 

Why do a -0.5 to get into the -0.5-0.5 range? The existing method does a -0.47 and scale via a factor 0.53/0.5. You said it's tweaking, but I would like to see a reference on what is the correct approach. In other words: where is that book you're talking about? :) Why not just do -0.47 (without scaling), for example? That seems a lot more logical and keeps the whites, greys and blacks a pure Y affair.

**Comment by sdsnatcher** [comment]
Why do I have this slight impression that you're wishing for pure whites
and blacks?   :)

The books I'm reading are the datasheets and all the references stated in
the spreadsheet, plus some service manuals to complement, like the HB-8000
and the VG-8020.

The page 2-17 of the TMS9918 datasheet clearly states 0.47V, and not 0.50V,
for both chroma values for both white and black. Texas engineers even
tweaked the scale it as much as they could by cutting out 0.3V at the top
of the scale, thus limiting the range to the typical 1Vpp stated in the
page5-3 of the same datasheet, and that is standardized by the TV-Encoders.
(check their respective datasheets).

0.47V is the output of the DAC value "14" and is 6% lower than the 0.50V
that is the middle of the 1Vpp scale (the would-be real neutral
white/black). As it's shown in the spreadsheet, the resolution of the DAC
is 0.03V, and the value 15 outputs _exactly_ 0.50V in the TMS9928/29 DAC.
Thus, if the Texas Engineers _really_ wanted pure whites and blacks they
could have easily done it, couldn't they? For reasons beyond my knowledge
they decided to keep the picture a bit warmer by setting up both the white
and the black with their pair of chroma values (R-Y and B-Y) to 14. You may
not like it, but that's how it's set up inside the chip.

The only precise way to boost that 0.47V to 0.50V in the analog circuitry
would to artificially add 0.03V to the whole scale. And, yes, this is a
form of tweaking. And that wouldn't be exactly cheap to do with 80's
discrete components.

The Philips VG-8020 seems to try to do a cheaper trick: it boosts the
entire R-Y scale by a percentage by using a lighter pull-down resistor
attached to the R-Y output of the TMS9929. The page 5-9 of the TMS9918
datasheet clearly states that the three outputs should have a 470 Ohms
pull-down resistor and a 15pF capacitor in parallel to ground, and Sharp
followed this by-the-book for the HB-8000 computer. The VG-8020, OTOH, has
3K3 resistors for Y and B-Y, and a 1K8 resistor for the B-Y.

This tweak of the VG-8020 will certainly affect the palette.  As I already
mentioned, many other MSX models will probably have such design deviations.
It's very easy to mess things up in YPbPr.  This is why I advocated that
the openMSX implementation should follow the TMS9918 datasheet recipe
by-the-book too.  And, by-the-book, the scale goes from 0V to 1V, and 0.47V
doesn't sit in the middle and will not produce pure blacks and whites
without tweaking.

The LM1889 TV encoder was recommended by the Texas in the page 3-6 of the
TMS9918 datasheet, and Sharp also followed this by-the-book for the
HB-8000.

Between the TMS9928 output and the LM1889, Sharp have placed a coupling
capacitor. This centers the signal  to -0.5V to +0.5V. Then Sharp adds 6V
to both R-Y and B-Y, and feed this same 6V floor as a reference to pin-3.
This is required because, as stated in the LM1881 datasheet page-8, the
angle phase of the subcarrier will be modulated by this value being
positive or negative when compared to the reference voltage on pin-3.
Here's where the reduced saturation occurs: the LM1881 expects the input
voltage to be in the -1V to +1V range relative to pin-3 (stated in that
same paragraph), but the circuit feds it with R-Y/B-Y in the 6V+/-0.5V
range. This explains why the HB-8000 has exactly 50% color saturation, but
without any tweaking in the tint.

Other MSX models with TMS9x2x VDPs and TV-Encoders also feature a similar a
level of saturation and provably where designed with similar
characteristics.

Yashica has also used coupling capacitors for the chroma signals in the
Philips VG-8020, except that the scale won't be -0.5V to +0.5V because of
the lighter pull-downs I already mentioned. It isn't possible to check how
this will affect the saturation because there's no complete datasheet of
the LVA510 available anywhere. I only have a partial 3rd party datasheet
here that doesn't contain this information.

​Was that enough information to convince you that the 0.47V wouldn't get
you pure whites and blacks without tweaking? I'm getting kinda exhausted of
having to explain the same thing again and again, you know.    :)

**Comment by MBilderbeek** [comment]
OK, I implemented the stuff :)

This is what I get for default (both saturations on 54%) - i.e. almost like the Wikipedia palette, but now without the tweaking:
![54-54](https://cloud.githubusercontent.com/assets/12675978/15793601/6251adac-29e5-11e6-9124-7ed5f614cc77.png)
This is with both on 100%:
![100-100](https://cloud.githubusercontent.com/assets/12675978/15793624/8eddf86c-29e5-11e6-95f5-8419b1786d66.png)
This is with 95%/95%:
![95-95](https://cloud.githubusercontent.com/assets/12675978/15793642/b28b985a-29e5-11e6-8ba1-5eeedf614ff6.png)
This is 50%/50%:
![50-50](https://cloud.githubusercontent.com/assets/12675978/15793668/d6fa4574-29e5-11e6-918a-c79c2d99a4a2.png)
and here is 50%/75%(Pb/Pr):
![50-75](https://cloud.githubusercontent.com/assets/12675978/15793709/13841330-29e6-11e6-8bf4-6c24e2df48a2.png)

Is this what you expect?

This is an outline of the code I used:

```
const float TMS99XXA_ANALOG_OUTPUT[16][3] = {
//          Y   R-Y    B-Y  voltages
    { 0.00, 0.47, 0.47 },
    { 0.00, 0.47, 0.47 },
    { 0.53, 0.07, 0.20 },
    { 0.67, 0.17, 0.27 },
    { 0.40, 0.40, 1.00 },
    { 0.53, 0.43, 0.93 },
    { 0.47, 0.83, 0.30 },
    { 0.73, 0.00, 0.70 },
    { 0.53, 0.93, 0.27 },
    { 0.67, 0.93, 0.27 },
    { 0.73, 0.57, 0.07 },
    { 0.80, 0.57, 0.17 },
    { 0.47, 0.13, 0.23 },
    { 0.53, 0.73, 0.67 },
    { 0.80, 0.47, 0.47 },
    { 1.00, 0.47, 0.47 },
};

const std::array<std::array<uint8_t,3>,16> VDP::getMSX1Palette() const
{
    assert(isMSX1VDP());
    if ((version & VM_TOSHIBA_PALETTE) != 0) {
        return TOSHIBA_PALETTE;
    } else {
        std::array<std::array<uint8_t,3>,16> tmsPalette;
        for (int color = 0; color < 16; color++) {
            // convert from analog output to YPbPr
            float Y  = TMS99XXA_ANALOG_OUTPUT[color][0];
            float Pr = TMS99XXA_ANALOG_OUTPUT[color][1] - 0.5f;
            float Pb = TMS99XXA_ANALOG_OUTPUT[color][2] - 0.5f;
            // apply the saturation
            Pr *= (saturationPr/100.0f);
            Pb *= (saturationPb/100.0f);
            // convert to RGB as follows:
            /*
              |R|   | 1  0      1.402 |   |Y |
              |G| = | 1 -0.344 -0.714 | x |Pb|
              |B|   | 1  1.722  0     |   |Pr|
            */
            float R = Y + 0.000f * Pb + 1.402f * Pr;
            float G = Y - 0.344f * Pb - 0.714f * Pr;
            float B = Y + 1.722f * Pb + 0.000f * Pr;
            // blow up with factor of 255
            R *= 255;
            G *= 255;
            B *= 255;
            // the final result is that these values
            // are clipped in the [0:255] range.
            tmsPalette[color][0] = Math::clipIntToByte(std::round(R));
            tmsPalette[color][1] = Math::clipIntToByte(std::round(G));
            tmsPalette[color][2] = Math::clipIntToByte(std::round(B));
        }
        return tmsPalette;
    }
}

```

**Comment by sdsnatcher** [comment]
The formulas seem correct. The only difference is that you do the scaling at the final step.

I checked the colors at 50%/50% saturation, and the values are a bit different from the spreadsheet. Out of the scanlines, your light red (color 9) show up as (250,151,130) on your image, and as (248,142,119) in the spreadsheet. I think this is caused by the openMSX's gamma correction, right?

If the colors from your image with both the gamma correction and the scanlines disabled match the colors of the spreadsheet, then you have nailed it.

Another interesting aspect no note: the screenshots you posted, even being png, seem to have a lot of subtle fluctuation in the color values inside each line, almost like there was a lossless compression in place. Is that correct?

**Comment by MBilderbeek** [comment]
Hahah, sorry, I used some settings in the postprocessor that caused all of this. I used a gamma correction of 1.1 and scanlines, blur and some noise. I should have posted a clean image... here it is again for 50/50:
![50-50](https://cloud.githubusercontent.com/assets/12675978/15800506/07c0901e-2a7c-11e6-9fce-a18b1ac8133e.png)
This is the palette I got according to the logging I made (not shown above in the code snippet):

```
0 0 4 0
1 0 4 0
2 58 187 69
3 112 211 120
4 84 89 212
5 123 123 230
6 179 99 76
7 97 223 230
8 212 106 85
9 248 142 120
10 199 199 92
11 217 212 132
12 54 165 61
13 176 107 172
14 199 208 197
15 250 255 248
```

So, slightly different than your spreadsheet....

**Comment by sdsnatcher** [comment]
Very good!

The slight differences might come for either, or a combination, of these two factors:
- In the spreadsheet, I first scale the values to the ranges required by the YPbPr conversion formula. In your implementation, scaling is done as the last step
- Differences in rounding in Excel versus C++

I'll leave it to you to decide if you want to adjust it or not, but I can say that I'm very satisfied with the results as they are. As we used the analog conversion formulas, we must remember that the real analog conversions will also show some variations like that.

Congratulations again to you and the openMSX team for such dedication to achieve the best possible emulation!

**Comment by MBilderbeek** [comment]
Here's a screenshot of a T6950 in the Toshiba HX-20E, which has its output directly connected to composite out. The saturation of colors 9 and 12 seems to be much higher than what we emulate now for the T6950.... @sdsnatcher any explanation for that?

![IMG_20190515_001115](https://user-images.githubusercontent.com/12675978/57755726-e6a0bb00-76f1-11e9-8873-cb9c082cfde0.jpg)

**Comment by friguron** [comment]
Hi, the pic is from my Toshiba HX-20E :)

I have to say that this red9 and green12 super saturation is present no matter the type of display: CRT or LCD...
That screenshot is taken from a SAMSUNG modern LCD, which (as usual also in modern times), has no modified controls from 50/50 positions (bright, contrast, saturation)... Tipically, no one plays with these controls anymore, as we did in the past...

My eternal doubt when using a CRT in the past was I somehow had my bright/contrast/saturation analog controls out of tune, and the saturation was caused by this... But apparently not, the high saturation is transmitted through the RF cable no matter what, so even a LCD modern display shows it.

My theory is that the "culprit" is the last element in the "Video out" circuit, which apparently seems to be the small metallic box (some RF modulator?) found in the additional PCB which holds the A/V out connectors. Yes well, maybe the palette encoded inside the VDP pin 37 (Composite OUT), is the one emulated inside openMSX, but being purists maybe some measurement should be done just before the signal enters the CRT (after that RF modulator circuit). And then, only then, we might discover then that the actual transmitted palette (and the one we should emulate) is the one recreated/distorted as in this last pic... (Just my wild guess). 
How to properly emulate this? No idea... Visually? fine tuning values just with test and error? There are not many ways to do it apart from that...

OTOH, I have no way of measuring this signals with my HX-20E though, no equipment, no probes, no oscilloscopes, etc... Sorry.

**Comment by friguron** [comment]
BTW, I've been reading some more infos around, and I've discovered that some SONY MSX-1's have this VDP also (!!!!!), specially the one I played more, from a friend of mine, a SONY HB-501P (Such an amazing beast!!) 

And guess what? I never knew this Sony and my Toshiba-HX-20E shared the VDP, and why? Because as far as I can remember, his red9 and green12 were much more in tune with MSX standard palette (!!). Well, I have to say this is now a blurry memory, but seriously... 

I can remember Salamander intro faces on his computer being much more "natural skin coloured" than with my Toshiba HX-20E (which were FIERY ferrari red, really irreal). If both machines shared the very same exact VDP, then it does mean that the RF modulators might have been altering this color info right before sending it onto the CRT. This might mean that, depending on the RF modulator, we might need to transform emulated palettes with an extra additional calculation...

Again, just wild guessing

**Comment by MBilderbeek** [comment]
@friguron Are you saying the screenshot was taken from the RF output and not from the composite video output?

**Comment by friguron** [comment]
@MBilderbeek Yep, but I can remember both outputs (RF and RCA video out (yellow)) yielding the same bleeding palette... I Will try to check it today if I can, as I'm not sure I have a proper RCA yellow video input TV at home right now (to compare both HX-20E outputs, RF and Composite Video)

**Comment by friguron** [comment]
@MBilderbeek, I've uploaded another pic of my T6950 but now using Composite Video (RCA Yellow cable), colors look exactly the same as in RF signal. Look:
 
![IMG_20190515_215813](https://user-images.githubusercontent.com/3338420/57812134-75521e00-776c-11e9-87f0-1f010f21ea2e.jpg)

OTOH, There are tons of artifacts, but these are caused by my Samsung LCD (which apparently becomes crazy with such a "regular and sharp" signal).

**Comment by friguron** [comment]
BTW, @MBilderbeek (and the rest of people), I've been pushing the limits of my toshiba-hx20E and its VDP. How? Running "modern demos", like the ones from dvik and so... 

For example, let's take unleashed.rom (rom version, ascii 16 mapper, run with sofarun under gr8net).
I invite you to run it with your TMS99XX MSX-1's (if you have one, I don't)... I have detected "bad behaviour" with my T6950 under unleashed.rom

There are 2 parts which look bad:

1.- Banner with 5 multicolor bars swirling around, trying to get in sync with border color at the same height... My NMS 8250 and openMSX both show perfect sync (yes, I know they're a V9938). OTOH,  my T6950 shows the border bars 1 or 2 pixels above where they should appear. Something like this:

![IMG_20190516_010507](https://user-images.githubusercontent.com/3338420/57815387-82283f00-7777-11e9-8e62-2f389a932be2.jpg)

2.- There's some part where there's a border scroller changing border color real quick in the BOTTOM border zone of the screen (beyond line 192). In my NMS 8250, all looks good, but in my T6950 the letters forming the scroll look skewed... Look:

![IMG_20190516_010719](https://user-images.githubusercontent.com/3338420/57815397-91a78800-7777-11e9-8e6c-bcc45c0c57e8.jpg)
I think this is a VERY INTERESTING and strange behaviour. I don't know if it should be emulable, or even explainable, but it's really interesting nonetheless...

I'm open to further testing...

**Comment by sdsnatcher** [comment]
> @sdsnatcher any explanation for that?

Remember that we have taken the color palete from the T7937A, because this one has RGB output. Given the lack of info on the previous Toshiba VDPs, that palette was used for the T6950(A) too.

But I have a Sony HB-20P and the colors seem very close to those of the T7937A, except that the CVBS color encoding tends to make the reds look more orangish.  So I can imagine two possible explanations

1) The HX-20 has a T6950 (without A), and this VDP has a different palette. Later the palette was made less saturated on the T6950A and T7947(A)

2) Toshiba might have messed up on the analog circuitry of the HX-20

There could have been a 3rd possibility (different palette from NTSC and PAL), but I already discarded this: I added an NTSC/PAL switch to my HB-20 to be able to do tests on this VDP on those two TV systems using a single machine.

There's are some factors to take into consideration too:

1) The saturation control on the Samsung LCD might be too high. They tend to setup it like this to make the image look more "vibrant". Try to enable a game mode on the TV and check the saturation control level (should be at ~ 50%). Test with other machines (an MSX2) via CVBS (not RGB) to see if the colors don't get too saturated too.

2) Smartphone cameras are pretty poor for this test. They tend to whitebalance the image and do some "saturation beautifying" behind the scenes. The best test would be to use a TV-Grabber card on a computer

I attached a picture of my Sony HB -20P as a comparison. Notice that at the color-5, the smartphone camera tried its tricks and created a mess. The colors also look more saturated on this photo than to the naked eye.



![IMG_5801](https://user-images.githubusercontent.com/13080763/57820125-d0cade80-7761-11e9-80f7-d28108f00660.JPG)

If you want to have an idea of how much the phone camera fiddled with the colors, this photo was taken under a cool white CFL. Notice that there's an area on the background where I marked "this should be white" that is horribly yellowed.

**Comment by friguron** [comment]
Hi, it's very tempting blaming the different (potentially) distorting steps taken from my CRT to the pics uploaded onto this github thread, but as I said before let's more or less debunk them one by one:

1.- As I said before, Samsung LCD controls are 50% saturated, and no silly "enhancer" modes are active. I know the famous Samsung trend to make images vibrant and saturated, but precisely these are the settings I disable as soon as I start using my samsung devices (TV's, mobile phones). In this case, no saturation controls are vibrant, neither outside 50%.

2.- Yesterday I uploaded a second pic from my Video Composite and colors look exactly the same as the pic Manuel uploaded from my RF output.

3.- This bleeding T6950 palette has been happening for decades with my Toshiba-HX20E, I mean, today I have this Samsung LCD, yes, but I also have a small 14" daewoo CRT, and both devices show these saturated colors... And even in the past, it also happened with my 14" SONY Trinitron monitor... And also with a previous ITT 19" TV, etc... I mean, my bad palette has been one of the many concerns of my "dreaded" HX-20E :(

Anyway, tonight I'll do the "definitive" test: I'll move my 2 MSX's onto my current Daewoo 14" CRT (also with saturation controls adjusted into a healthy 50%). Once there, we'll compare the "perfect" V9938 palette from my NMS 8250 against the ugly T6950 one, with a video... I have in mind testing the same image. I'll start with the colorbars.bas program, and aftewards we'll test the Toshiba HX-20 city landscape boot screen on both computers (It's very funny running the Toshiba WP-Rom in the NMS-8250 :)

Of course we can expect some distortion of colors from reality, but believe me when I say that, luckily, my Samsung mobile gets a REALLY ACCURATE palette from my displays (I confirm that's a rare and happy case when it' happening).
OTOH, white balance with that pics is more than ok, as I take my pics in complete darkness...
Even Samsung mobile phone might distort colors, I tell you all colors execpt red9 and green12 are normally saturated on the display... And this is what the mobile phone shows also... I mean, even we might not get the pure RGB colors in the mobile phone display, the contrast between all colors being normally saturated except red9 and green12 is somewhat proof enough...
T6950/HX-20E problems are not hue, but saturation related.

But again, enough rambling, tonight I'll make a comparison test between my NMS-8250 vs Toshiba HX-20, against the very same (quite decent) CRT screen, with carefully 50% saturated controls, with a video. Let's hope we all get what my eyes sees :) And if not, at least by comparison, you'll get my point.

Greetings.

**Comment by sdsnatcher** [comment]
Hey, calm down. the monitor/cell phone was just one of the possible factors that I explained. Take a look at the others too, plz.

Does the HX-20 have the T6950 or T6950A inside? Is the VDP socketed?

Do you have any machine with the TMS9x18 VDP around? That would be a good comparison too, since these two VDPs also feature an internal CVBS encoder.
Please note that it *cannot* be a TMS9x28 or TMS9x29. These use external TV-encoder chips.

**Comment by friguron** [comment]
You can browse the "ripping" on demand I did 3 days ago together with @MBilderbeek here:
https://drive.google.com/drive/folders/1XwigXkw5hxYi1kuQGCvasXeWVWM_cW8f

Unluckily I don't have any PAL TMS9X18 MSX-1 vdp around... I only have my V9938 and my T9650 (not A)...

The VDP is socketed, yes, but its pinout is not compatible with TMS9x18 (42 vs 40 pins... and some other peculiarities) In the case I might get some TMS original VDP, I should add some hardware modification to be able to use it against this socket...

I'm as clueless as you all... :( 
The saturation issue of my Toshiba equipment has been bugging me all life and "I want a solution" also (if ever, I doubt there is one :) I want to make consistent and sistematic tests to see how we can pin down the source of this saturation...

In the worst case, we should/might add this palette distortion to openmsx if we find it consistent among all HX-20 T6950's around... 
If we could get some other users with this Toshiba-PAL-VDP combo...

Greetings

**Comment by friguron** [comment]
BTW, someone already made a rendition of T6950 palette in this post: http://bifi.msxnet.org/paledit/?t=2&d=palette
Apparently it's a "visually modelled" palette (by hand), but anyway we can notice how this other user also considers he sees heavily saturated red9 and green12 colors... 

I'm intrigued now if revisions A and B of T6950 VDP might have corrected their internal palettes to a more MSX-1 "official" one... But a quick query on internet for getting one of these IC's yield near to 0 results... 

That's the point nº 1 of your theory... If I ever got a T6950**A** or **B** I would replace it on my VDP socket without any problem for further testing.

**Comment by sdsnatcher** [comment]
> Unluckily I don't have any PAL TMS9X18 MSX-1 vdp around... I only have my V9938 and my T9650 (not A)...
> 
The TMS9x18 are NTSC only.


> The VDP is socketed, yes, but its pinout is not compatible with TMS9x18 (42 vs 40 pins... and some other peculiarities) In the case I might get some TMS original VDP, I should add some hardware modification to be able to use it against this socket...
> 

If you want to investigate yourself this topic further, UtSource has the T6950A on sale:
https://www.utsource.net/itm/p/365317.html

You’ll then be able to swap between the two versions to compare what has changed between the two, including the Palette. Then you can open a new issue and detail to the openMSX team what needs to be implemented.

Sadly, I won’t have free time to help you in this journey. You can check the history of this issue how it was done for the T7937A.

**Comment by friguron** [comment]
As I said before, I've uploaded 1 more video and image to compare my 2 machines palette with the very same CRT, **at the same time**, (HX-20E T6950 vs NMS 8250 V9938). I know, it's not the proper comparison to perform, but you might finally get how saturated the image is, using the "relative comparison method", against a proper palette with clear red9 color and dark green12 (Recommended resolution to watch the video: 1080 HD)

Important note: my Philips NMS 8250 is running the toshibahx20.rom bios so it can show the very same image as the real Toshiba (which is just idly booting). The video seems to be alternating between RF and AV (and it is!), but I've connected 2 different machines on each of the 2 connections (!): RF -> Toshiba, AV -> 8250
It's fun how they're running the same software: The 8250 an "alien" toshiba rom via GR8NET, the Toshiba just booting its own brain :)

I've also uploaded an image to check and confirm that the actual color/bright/contrast/sharpness controls from my CRT were indeed far from saturation.

You can find those 2 documents, (as well as all other multimedia stuff from the last days related with my HX-20E), here: https://drive.google.com/drive/folders/1XwigXkw5hxYi1kuQGCvasXeWVWM_cW8f 
These should be the last image uploads I'll be adding regarding my HX-20E. I think everything had been said and tested.

BTW, @sdsnatcher don't worry, I don't think I will be ordering a T6950-A variant in the near future... I don't have neither the expertise nor the hardware (nor that much free time either!) to test my VDP's as thorougly as it was apparently done with the T7937A some time ago...

OTOH, I would say this long thread has also turned into some form of topic hijacking many messages ago (sorry), so, unless some groundbreaking discovery happens, and/or someone finds out any new relevant data, we'd better stop adding more offtopic noise to the thread.
As you said, in case someone gets new info we should **then** open a new thread, with the topic being just the T6950 bleeding palette and nothing more.

That should be all.

It's been a pleasure discovering so many things with you all, greetings.

**Comment by MBilderbeek** [comment]
The artifacts in that MSX unleashed demo are probably explained by the fact that this machine has a separate crystal for the VDP (Q2) and CPU (Q1).

**Comment by MBilderbeek** [comment]
Yes, other MSX computers will also show it, when they have separate xtals. I believe the Philips NMS 8280 will also show glitches, as it has a slightly different timing due to the genlocker. (I think you can read something about it in the readme of the IO demo.)

**Comment by npiegdon** [comment]
I'm about six years late to this discussion, but I might have an interesting detail to add.  At the risk of complicating things a little further, I'm not convinced the output of the TMS9x2x is intended to be YPbPr.  There is an old TI "Application Report" document called the [TMS9928/29 and TMS9128/29 Interface to Color Monitors](http://spatula-city.org/~im14u2c/vdp-99xx/e3/SPPA004A_9928-29_9118-28_Interface_to_Monitors.pdf) that seems to have only resurfaced in the last handful of years.

It describes a circuit that can be used to adapt the VDP's output to RGB (with the caveat that "the purpose of this circuit is to produce a bright, sharp image from the TMS9928 rather than to produce a very accurate color difference to the R-G-B converter...".

That said, in the body of the text (near the beginning of the "R-G-B Converter" section), there is the line "... in conformance to standard industry practice, the R-Y output of the video display processors has been divided by a factor of 1.14, and the B-Y by a factor of 2.03."

Looking at most conversion references that make the distinction (like [this one](https://web.archive.org/web/20180421030430/http://www.equasys.de/colorconversion.html)), 1.14x and 2.03x are the factors for a YUV to RGB conversion.  YPbPr to RGB uses 1.402x and 1.772x respectively, which is a fair bit different.

It makes me wonder if those lines on the chip shouldn't have been labeled U and V instead of B-Y and R-Y?

I was still in the information gathering stages when I ran across this mostly-current discussion (relative to, say, Richard Drushel's articles from the previous century), so I haven't had a chance to run the numbers (using sdsnatcher's wonderful spreadsheet).  While I've found lots of discussion [here-and-there](https://atariage.com/forums/topic/138149-tms-9918a-and-992xa-color-differences/) about the "true" palette values, I've never seen those discussions connected to the numbers given by TI themselves in that application report.

I wonder how those palettes would look using the YUV-to-RGB conversion instead?

**Comment by briantw** [comment]
YPbPr is a specific colour space, and the TMS9929ANL et al were not designed around that colour space. The TMS9929 (and TMS9928) outputs Y, R-Y, and B-Y, but the colour space is not identical to YPbPr, and if you're expecting YPbPr, the colours will be off.

**Comment by npiegdon** [comment]
@briantw Were you replying to me or the top-post?  I wasn't expecting YPbPr.  It's definitively YUV; specifically using the (fairly standard) matrix at the [equasys site](https://web.archive.org/web/20180421030430/http://www.equasys.de/colorconversion.html).

Since my reply a year ago, we (Ikrananka at the AtariAge forums and myself) have combed through a bunch of original references--internal TI memos, conversations between the US and EU TI teams, and some application notes--and gathered some of our own vectorscope data (with [Artemio Urbina](https://junkerhq.net/)'s help) to compare the 992xA to the composite output from the 9918A, all of which pointed to YUV being the right answer beyond any shadow of a doubt.

Here's the vectorscope shot of the 9918A with the various hypothetical 992xA conversions overlaid:

![comparison](https://user-images.githubusercontent.com/2000287/146625516-94186aa7-bb3c-484a-85b3-f884fbe5e8b7.jpg)

Neither the "plain" R-Y/B-Y color difference version nor the YPbPr conversion works as well as YUV.

The only quirk is that--at least in the Colecovision--the NTSC/PAL video output is a little hot, so a conversion to RGB clips a couple of the colors.
